src/pwm.rs

```diff
@@ -644,45 +831,286 @@ macro_rules! tim_hal {
 
                 unsafe { MaybeUninit::<PINS::Channel>::uninit().assume_init() }
             }
+
+            impl PwmAdvExt for $TIMX {
+                type Rec = rec::$Rec;
+
+                fn pwm_advanced<PINS, CHANNEL, COMP>(
+                    self,
+                    _pins: PINS,
+                    prec: Self::Rec,
+                    clocks: &CoreClocks,
+                ) -> PwmBuilder<Self, PINS, CHANNEL, FaultDisabled, COMP, AlignmentLeft>
+                where
+                    PINS: Pins<Self, CHANNEL, COMP>
+                {
+                    prec.enable().reset();
+
+                    let clk = $TIMX::get_clk(clocks)
+                        .expect("Timer input clock not running!")
+                        .0;
+
+                    // Write prescale 0
+                    self.psc.write(|w| w.psc().bits(0));
+
+                    // Set TOP value to max
+                    self.arr.write(|w| w.arr().bits(<$typ>::MAX));
+
+                    PwmBuilder {
+                        _tim: PhantomData,
+                        _pins: PhantomData,
+                        _channel: PhantomData,
+                        _fault: PhantomData,
+                        _comp: PhantomData,
+                        _alignment: PhantomData,
+                        base_freq: clk.hz(),
+                    }
+                }
+            }
+
+            impl<PINS, CHANNEL, FAULT, COMP, ALIGNMENT>
+                PwmBuilder<$TIMX, PINS, CHANNEL, FAULT, COMP, ALIGNMENT>
+            where
+                PINS: Pins<$TIMX, CHANNEL, COMP>,
+            {
+                pub fn finalize(self) -> (PwmControl<$TIMX, FAULT>, PINS::Channel) {
+                    let tim = unsafe { &*$TIMX::ptr() };
+
+                    $(
+                        // BDTR: Advanced-control timers
+                        // Set CCxP = OCxREF / CCxNP = !OCxREF
+                        // Refer to RM0433 Rev 6 - Table 324.
+                        tim.$bdtr.modify(|_, w| w.moe().$moe_set());
+                    )*
+
+                    tim.cr1.modify(|_, w| w.cen().enabled());
+
+                    unsafe {
+                        MaybeUninit::<(PwmControl<$TIMX, FAULT>, PINS::Channel)>::uninit()
+                            .assume_init()
+                    }
+                }
+
+                /// Set the PWM frequency; will overwrite the previous prescaler and period
+                pub fn frequency<T: Into<Hertz>>(self, freq: T) -> Self {
+                    let tim = unsafe { &*$TIMX::ptr() };
+
+                    let freq: Hertz = freq.into();
+                    let reload: u32 = self.base_freq.0 / freq.0; // u32
```

Comment:
For example: 1000 / 92 = 10; -> 100Hz frequency

You can make the it less than or equal to the frequency specified with this

```let reload: u32 = (self.base_freq.0 + freq.0 - 1) / freq.0; // u32```

Then: 1091 / 92 = 11; -> 90.9Hz frequency

Reply:
Is there a clean way to use a builder pattern with typestate programming while doing all the configuration in the finalize method?

I would like it if I could run some conditional code (if or match) on the type of some of the type states; if we're left-aligned, use one bit pattern, otherwise if we're center-aligned, use a different bit pattern and adjust the period to match. But I'd rather not define a trait for each parameter of the struct and implement those traits separately for each value of each type parameter.

I was going to say there isn't much inter-dependence between the sub-parameters, but I just realized that .frequency().center_aligned() will act differently than .center_aligned().frequency(), so at least the alignment really should be handled in the finalize method. Is there a better way to do it than just adding a new trait for alignment that sets those bits properly?

For rounding, usually the error will be a fraction of a percent, but if you're doing a high frequency digital power topology and using a low base clock to reduce power consumption, I would want it to round to nearest instead of up or down:

```
let reload: u32 = (self.base_freq.0 + freq.0 >> 1) / freq.0; // u32
```

If you don't have any examples of running code differently inside a function based on type parameters, I'll just add a new trait for alignment to handle that at finalize time similar to the PwmPinEnable trait I added to handle enabling/disabling channels properly based on the complementary state. I feel like period and prescaler can still be handled the way they are now, at least until const generics is stabilized.

Reply:
>Is there a clean way to use a builder pattern with typestate programming while doing all the configuration in the finalize method?

Hmm I don't think I've seen typestates used on a builder struct before, so I don't have any examples to compare against. Some use of typestates is quite nice here (for example `FaultEnabled`/`FaultDisabled`), but my first suggestion would still be to save the alignment and frequency values as members of `PwmBuilder` and then apply them at finalize time.

> I would want it to round to nearest instead of up or down

Sounds good to me! I would just mention that in a doc comment also.